Add module-info descriptors and ResourcesProvider

jperedadnr opened the pull request:
<!--- Provide a brief summary of the PR -->

### Issue

<!--- The issue this PR addresses -->
Fixes #719

### Progress

<!-- Please ensure you actioned and ticked each box below before requesting a review -->

- [x] Change must not contain extraneous whitespace
- [x] License header year is updated, if required
- [x] The PR name must follow the [pre-defined format](https://github.com/gluonhq/scenebuilder/blob/master/CONTRIBUTING.md)
- [x] Verify the contributor has signed [Gluon Individual Contributor License Agreement (CLA)](https://docs.google.com/forms/d/16aoFTmzs8lZTfiyrEm8YgMqMYaGQl0J8wA0VJE2LCCY)

johanvos commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/i18n/spi/I18NResourcesProviderImpl.java:
```diff
@@ -29,26 +29,32 @@
-        String resourceName = toResourceName(bundleName, "properties");
-        try (InputStream is = loader.getResourceAsStream(resourceName);
-             InputStreamReader isr = new InputStreamReader(is, "UTF-8");
+        String resourceName = ResourceBundle.Control
```
If possible, add the resourceName to the printed exception (L61) in case an exception occurs. That way, users might be able to easily detect what is wrong (or provide better feedback)

jperedadnr replied:
done

Oliver-Loeffler commented:
That was a lot of work. I'll finish the review tonight.

abhinayagarwal reviewed (approved):
Tested on both vanilla maven app and bundled dmg.

Oliver-Loeffler reviewed (changes requested):
Technically this change is fine when running plain Maven based builds. Interestingly, this setup generates errors when using Eclipse and importing SceneBuilder as a Maven-Project.
Here an error is created which prevents project compilation as there is a test class in kit wich has no package declaration. Not sure if this is Maven's fault or Eclipse fault, it does not work out of the box in Eclipse.

![Project-Error](https://github.com/user-attachments/assets/31b860ad-61eb-41bd-a69c-e47ba8050f12)

Tests are different source trees, and in other projects I have made positive experience with separating tests into a different sub-project. 

The error does not occur in Idea 2024.2.0.1, but indeed in Eclipse 2024-09 (2024-09, 4.33.0, Build id: 20240905-0614).

What do you think about this?

jperedadnr commented:
Thanks for your review, @Oliver-Loeffler 

I've added the `require transitive`.

As for `InaccessibleObjectException`, I've added to the pom:

```
               <plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-surefire-plugin</artifactId>
                    <version>3.0.0-M5</version>
                    <configuration>
                        <reuseForks>false</reuseForks>
                        <forkCount>1</forkCount>
 +                       <argLine>--add-opens=javafx.fxml/javafx.fxml=com.gluonhq.scenebuilder.kit</argLine>
                    </configuration>
                </plugin>
```
So I guess that you need the same if you run tests from within your IDE and not from the plugin?

As for the failing test on Eclipse, it kind of makes sense, under the modular world a class has to have a package name, so even if you move it to a third maven module "tests", the issue will remain the same unless, of course, we have non-modular-tests maven module (after all, we would be testing issues from any JavaFX project using FXML files, but not modules), and also modular-tests maven module.

I'm fine with moving tests to its own module(s) (in fact, I'm back working on "Components", after #241 ), but this should be done in a follow-up PR. So maybe we need to comment this test out (and related classes).

Oliver-Loeffler commented:
@jperedadnr I've reviewed a 2nd time and identified some more packages and modules which require the `transitive` keyword. I've also found 2 missing `exports`. After that, I thing the PR is ready to by merged. I agree with your proposal.

jperedadnr commented:
Can you list those changes? I don't find those issues from my IDE.

Oliver-Loeffler commented:
> Can you list those changes? I don't find those issues from my IDE.

## Changes in kit:

So following changes in `kit` `module-info.java`:

```java
    requires transitive javafx.fxml;
    requires transitive javafx.web;
    requires transitive static java.prefs;
    requires transitive static aether.api;
```

Classes to become public:

* `com.oracle.javafx.scenebuilder.kit.skeleton.SkeletonContext`

## Changes in app:

`module-info.java`

```java
requires transitive com.gluonhq.scenebuilder.kit;
exports com.oracle.javafx.scenebuilder.app.preferences;
exports com.oracle.javafx.scenebuilder.app.menubar;
```

Classes to become public:

* `com.oracle.javafx.scenebuilder.app.ResourceController`

## Warnings in Eclipse IDE (partial listing):

* kit

![grafik](https://github.com/user-attachments/assets/9ae5b429-1d5d-474b-bf43-24cb8ab1fd0f)

* app

![grafik](https://github.com/user-attachments/assets/feb4d74e-4a7f-4528-a394-b2ef64973365)

I've added a review commend at the particular positions in the respective module-info files.

jperedadnr commented:
I've added transitive to the javafx and java.pref modules in kit, and transitive to kit module in app.

I haven't added transitive to aether.api though, as that means that anyone consuming the kit will need to add the aether module, or :
```
[ERROR] Compilation failure
[ERROR] module not found: aether.api
```

What error do you get without this? Or is it just a warning?

jperedadnr commented:
Note about Warnings/Errors: 
If I add to the maven compiler plugin:
```
<configuration>
                        <useIncrementalCompilation>false</useIncrementalCompilation>
                        <showWarnings>true</showWarnings>
                        <compilerArgs>
                            <arg>-Xlint:all</arg>
                            <arg>-Werror</arg>
                        </compilerArgs>
                    </configuration>
```
regardless the IDE, we'll get tons of warnings and the compilation will fail. I see many of those warnings related to the new modules, but do we really need to address those now?

For instance:
```
class com.oracle.javafx.scenebuilder.kit.util.PaintConvertUtil in exported package com.oracle.javafx.scenebuilder.kit.util declares no explicit constructors, thereby exposing a default constructor to clients of module com.gluonhq.scenebuilder.kit
```

easily fixed with 
```
    PaintConvertUtil() {
        // no-op
    }
```
Should we do that for all warnings (!!!) now?

Oliver-Loeffler commented:
I was not worrying about all these warnings. It was indeed just those where you now already have added the transitive keyword, so that one can work using the kit module. However, the `aether.api` item is similar to the `SkeletonConverter` / `SkeletonContext` public API situation. There are many classes which expose parts of `aether.api`. Having them not exported (or declared transitive) may limit the usage if kit.

We export many items of the kit maven package:

```java
    ...
    exports com.oracle.javafx.scenebuilder.kit.editor.panel.library.maven;
    exports com.oracle.javafx.scenebuilder.kit.editor.panel.library.maven.preset;
    exports com.oracle.javafx.scenebuilder.kit.editor.panel.library.maven.repository;
    exports com.oracle.javafx.scenebuilder.kit.editor.panel.library.maven.repository.dialog;
    exports com.oracle.javafx.scenebuilder.kit.editor.panel.library.maven.search;
    ...
```

And those items have public API refering to `aether.api` classes. Hence my proposal to declare `aether.api as transitive.
All the other warnings as you have mentioned, should be addressed in a separate PR.

jperedadnr commented:
Right, I took the shortcut of exporting every possible package from Kit (most of them are used by App), and because I didn't want to impose any limitation on third party projects using Kit.

Based on the lint warnings, I've added the default missing constructors, and also the `requires transitive` for aether.api.

There are still a few warnings (58 for Kit, 7 for App, but few of them related to modules now, about using automatic-modules)
About lint warnings, I've filed #725 and for some more urgent fix: #726

Oliver-Loeffler commented:
> There are still a few warnings (58 for Kit, 7 for App, but few of them related to modules now, about using automatic-modules) About lint warnings, I've filed #725 and for some more urgent fix: #726

I have also started a new issue #727 to address the problem with tests in Eclipse.
